### PR TITLE
Fetch and checkout pr branch on gitpod if LILA_PR variable is provided and get preview easily

### LINKED ISSUE #17: Lila PR preview sites
Support a link that, when clicked, will start a gitpod instance but with a PR's branch checked out instead of `master`. Will allow others to test a PR easily.

You can [setup environment variables on Gitpod](https://www.gitpod.io/docs/configure/projects/environment-variables#providing-one-time-environment-variables-via-url) in the url like this:

```diff
- https://gitpod.io/#https://github.com/lichess-org/lila-docker
+ https://gitpod.io/#lila-pr=12345/https://github.com/lichess-org/lila-docker
```

Then, if `lila-pr` is set, we can fetch and checkout during setup.

```bash
git fetch upstream pull/12345/head:preview
git checkout preview
```

### PR DESCRIPTION
**Pull request by Carbrex**
Closes #17
Try this
https://gitpod.io/new/#LILA_PR=14733/https://github.com/Carbrex/lila-docker/tree/patch-4

Works well on my local and also on gitpod. Any idea why the docker compose build test failed.
Also why doesn't these tests show up on my fork like how the lila's do.

### CODE REVIEW

**Comment by fitztrev** [comment]
> Also why doesn't these tests show up on my fork like how the lila's do.

I believe you can enable them for your fork if you visit https://github.com/Carbrex/lila-docker/actions as the repo owner

Looks like the docker build is failing because of an issue with the mobile repo. Could be temporary or maybe the dependencies or setup changed. I'll investigate.

**Comment by fitztrev** [comment]
Testing it out now. This is great!

Feel free to push as many followup commits as you want. I'll squash merge once it's ready.

File: command/src/main.rs [hunk 1]
```diff
@@ -281,6 +283,49 @@ fn setup(mut config: Config) -> std::io::Result<()> {
...
+
+    let status = cmd.status().unwrap();
+    if !status.success() {
+        println!("Failed to fetch PR");
+    }
+
+    let mut cmd = std::process::Command::new("git");
+    cmd.current_dir("repos/lila")
+        .arg("checkout")
+        .arg(format!("pr-{pr_no}"));
```
Comment (fitztrev) [inline command/src/main.rs:323]: In the console logs, I see the stdout of `Switched to branch 'pr-14733'`, could we add a cliclack-formatted success message too?

File: command/src/main.rs [hunk 2]
```diff
@@ -281,6 +283,49 @@ fn setup(mut config: Config) -> std::io::Result<()> {
...
+
+    let status = cmd.status().unwrap();
+    if !status.success() {
+        println!("Failed to checkout PR branch");
+    }
```
Comment (fitztrev) [inline command/src/main.rs:328]: I would just pull this whole block code out of `setup()` into its own function, like `gitpod_checkout_pr()`

**Review by fitztrev — COMMENTED** [review]
I think when a PR is passed, we can skip the other prompts (optional services, passwords, etc) and go for an opinionated setup so that things just boot right away. In the future, we could control the optional services using the same technique if they're needed. Doesn't have to be a part of this PR, but I might try to do that if you don't want to.

**Comment by Carbrex** [comment]
> I think when a PR is passed, we can skip the other prompts (optional services, passwords, etc) and go for an opinionated setup so that things just boot right away. In the future, we could control the optional services using the same technique if they're needed. Doesn't have to be a part of this PR, but I might try to do that if you don't want to.

So, do I not add any optional services, when  the lila_pr variable is passed?

**Comment by Carbrex** [comment]
Should Cargo.lock be in .gitignore?

File: command/src/main.rs (hunk 1 above)
Comment (Carbrex) [reply command/src/main.rs]:
```
│
◇  Cloned lichess-org/lila ✓
│
◇  Cloned lichess-org/lila-ws ✓
│
◇  Cloned lichess-org/lila-db-seed ✓
│
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                remote: Enumerating objects: 607175, done.
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                remote: Counting objects: 100% (607172/607172), done.
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                remote: Compressing objects: 100% (135758/135758), done.
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                remote: Total 601293 (delta 415641), reused 592908 (delta 407337), pack-reused 0
Receiving objects: 100% (601293/601293), 325.62 MiB | 54.48 MiB/s, done.
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◓  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                
◒  Fetching  PR-14733 from lila                                                                                                                                
◐  Fetching  PR-14733 from lila                                                                                                                                Resolving deltas: 100% (415641/415641), completed with 4277 local objects.
◓  Fetching  PR-14733 from lila                                                                                                                                From https://github.com/lichess-org/lila
 * [new ref]               refs/pull/14733/head -> pr-14733
◓  Fetching  PR-14733 from lila                                                                                                                                 * [new tag]               pre-bcrypt           -> pre-bcrypt
 * [new tag]               pre-paypal-checkout  -> pre-paypal-checkout
 * [new tag]               pre-responsive       -> pre-responsive
 * [new tag]               v1.0.0               -> v1.0.0
 * [new tag]               v1.1.0               -> v1.1.0
◓  Fetching  PR-14733 from lila                                                                                                                                warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.scalachess.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.scalachess.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/tagmanager.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/tagmanager.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.submodules/pdfexporter.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.submodules/pdfexporter.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.submodules/boardcreator.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.submodules/boardcreator.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.ui/chessli.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.ui/chessli.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/moment.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/moment.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/ChessPursuit.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/ChessPursuit.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/multiple-select.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/multiple-select.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/hopscotch.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/hopscotch.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/shepherd.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/shepherd.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/stockfish.js.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/stockfish.js.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/flatpickr.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/flatpickr.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/bar-rating.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/bar-rating.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/visualizer.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/visualizer.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/Sunsetter8.path'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/Sunsetter8.url'. Skipping second one!
warning: b0e3a57405c9c1c62f81817f357701aacefe47a3:.gitmodules, multiple configurations found for 'submodule.public/vendor/Sunsetter8.branch'. Skipping second one!
◐  Fetching  PR-14733 from lila                                                                                                                                Could not access submodule 'modules/swisssystem' at commit 73a59c3166
Could not access submodule 'play' at commit 6451639122
Could not access submodule 'public/staunton' at commit 71a13aafb3
Could not access submodule 'public/vendor/Sunsetter' at commit 26220b5ccb
Could not access submodule 'public/vendor/Sunsetter' at commit 4a18768ff8
Could not access submodule 'public/vendor/bar-rating' at commit 5844a91da9
Could not access submodule 'public/vendor/flatpickr' at commit 405e90eab5
Could not access submodule 'public/vendor/hopscotch' at commit a45486db83
Could not access submodule 'public/vendor/jquery-textcomplete' at commit 235d5a1df9
Could not access submodule 'public/vendor/moment' at commit a0c69f5d13
Could not access submodule 'public/vendor/multiple-select' at commit 1164e71a89
Could not access submodule 'public/vendor/pgn4web' at commit b728ce663b
Could not access submodule 'public/vendor/shepherd' at commit c691d3042b
Could not access submodule 'public/vendor/stockfish.js' at commit 2a45de45e2
Could not access submodule 'public/vendor/stockfish.pexe' at commit 685a5d225f
Could not access submodule 'public/vendor/tagmanager' at commit 6ee267e01b
Could not access submodule 'public/vendor/timeago' at commit fa0f476ead
Could not access submodule 'public/visualizer' at commit 839b19f7a8
Could not access submodule 'modules/chess' at commit c4954c210c
Could not access submodule 'submodules/boardcreator' at commit 3958d1ddc2
Could not access submodule 'submodules/evaluator' at commit fbf2d396c6
Could not access submodule 'submodules/pdfexporter' at commit 071aad2d10
Could not access submodule 'ui/chessli' at commit 951d96b455
◇  Fetched PR ✓
│
Switched to branch 'pr-14733'
◇  Switched to branch 'pr-14733' ✓```

Comment (Carbrex) [reply command/src/main.rs]: It shows up like this multiple fetching pr  msgs and these warnings. How to fix them?

Comment (Carbrex) [reply command/src/main.rs]: Much better after https://github.com/lichess-org/lila-docker/pull/60/commits/5a7005e2da6a8153362149b16c13daea33547d29
```
◇  Cloned lichess-org/lila-ws ✓
│
◇  Cloned lichess-org/lila-db-seed ✓
│
◒  Fetching  PR-14733 from lila                                                                                                                                remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (14/14), done.
◐  Fetching  PR-14733 from lila                                                                                                                                remote: Total 15 (delta 13), reused 1 (delta 0), pack-reused 0
Unpacking objects: 100% (15/15), 1.09 KiB | 278.00 KiB/s, done.
From https://github.com/lichess-org/lila
 * [new ref]         refs/pull/14733/head -> pr-14733
◓  Fetching  PR-14733 from lila                                                                                                                                Fetching submodule public/vendor/ChessPursuit
◇  Fetched PR ✓
│
Switched to branch 'pr-14733'
◇  Switched to branch 'pr-14733' ✓
```

**Comment by Carbrex** [comment]
main.rs is now too big to be in a single file?

**Comment by fitztrev** [comment]
> Should Cargo.lock be in .gitignore?

Yes, you can commit it.

> main.rs is now too big to be in a single file?

I think it's OK. We'll ask @kraktus for a Rust review once it's ready, to be sure.

File: command/src/main.rs (hunk 1 above)
Comment (Carbrex) [reply command/src/main.rs]: Can you help with this how to remove these msgs from coming up
```
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (14/14), done.
```

**Comment by Carbrex** [comment]
Looks like the functionality broke after these changes.

**Comment by Carbrex** [comment]
Thanks for fixing it. I got it working on my patch-3 branch though. https://github.com/Carbrex/lila-docker/tree/patch-3  A separate lila_pr variable would help in making the other changes you mentioned above?

> I think when a PR is passed, we can skip the other prompts (optional services, passwords, etc) and go for an opinionated setup so that things just boot right away. In the future, we could control the optional services using the same technique if they're needed. Doesn't have to be a part of this PR, but I might try to do that if you don't want to.

**Comment by fitztrev** [comment]
Yep good point, how about we add both to the struct. keep `workspace_context` and add `lila_pr_no` as its own.

Didn't mean to step on your toes with the changes. I had some in-progress edits while reviewing when you had pushed. I'll pause on making any changes until you've finalized.

**Comment by Carbrex** [comment]
> Yep good point, how about we add both to the struct. keep `workspace_context` and add `lila_pr_no` as its own.
> 
> Didn't mean to step on your toes with the changes. I had some in-progress edits while reviewing when you had pushed. I'll pause on making any changes until you've finalized.

Ok, will complete within 10 minutes and then you can push your changes. Thanks for helping.

**Comment by Carbrex** [comment]
I have done my changes in patch-3 but I am having trouble merging my changes with yours.

File: command/src/main.rs (hunk 1 above)
Comment (fitztrev) [reply command/src/main.rs]: Update the progress output so only errors will show

**Comment by Carbrex** [comment]
Finally done. You can proceed now. Sorry for all this clutter.

**Review by fitztrev — APPROVED** [review]

**Comment by fitztrev** [comment]
tl;dr - Optional URL key/value will start a Gitpod workspace with a PR checked out instead of master.

Example: https://gitpod.io/new/#LILA_PR=14738/https://github.com/Carbrex/lila-docker/tree/patch-4

Those key/value pairs get automatically added to GITPOD_WORKSPACE_CONTEXT env so this PR parses and then runs the necessary git commands.

**Comment by kraktus** [comment]
I’ll check tomorrow 👍

**Comment by kraktus** [comment]
Nice work, only changed a few nits. The failure of test is not due to change, the tests are inherently flickered  because of setting env variables for their own process.

File: command/src/main.rs [hunk 3]
```diff
@@ -134,18 +135,48 @@ impl Repository {
...
+        let workspace_url =
+            std::env::var("GITPOD_WORKSPACE_URL").expect("Missing env GITPOD_WORKSPACE_URL");
+
+        let workspace_context: GitpodWorkspaceContext = serde_json::from_str(
+            &std::env::var("GITPOD_WORKSPACE_CONTEXT")
+                .expect("Missing env GITPOD_WORKSPACE_CONTEXT"),
+        )
+        .expect("Failed to parse GITPOD_WORKSPACE_CONTEXT");
+
+        let lila_pr_no = workspace_context.envvars.as_ref().and_then(|envvars| {
```
Comment (kraktus) [inline command/src/main.rs:168]: It's strange not to use `workspace_context` to get LILA_PR, because it currently make the former useless.

Comment (Carbrex) [reply command/src/main.rs:168]: It is using workspace_context, isn't it?

Comment (fitztrev) [reply command/src/main.rs]: Good point. Changed.

**Comment by fitztrev** [comment]
Great feature. Thanks @Carbrex! :tada: :rocket:

Thanks for the review @kraktus!
